File: python/year_2023/day_15.py
from pathlib import Path
from functools import reduce


def part_1(input: str) -> int:
    return sum(map(hash, input.split(",")))


def part_2(input: str) -> int:
    boxes = [{} for _ in range(256)]
    for operation in input.split(","):
        if operation[-1] == "-":
            label = operation[:-1]
            box_idx = hash(label)
            boxes[box_idx].pop(label, None)
        else:
            label, _, focal_length = operation.partition("=")
            box_idx = hash(label)
            boxes[box_idx][label] = int(focal_length)
    return sum(
        box_idx * slot * focal_length
        for box_idx, box in enumerate(boxes, 1)
        for slot, focal_length in enumerate(box.values(), 1)
    )


def hash(input: str) -> int:
    return reduce(op, map(ord, input), 0) % 256


def op(value: int, code: int) -> int:
    return (value + code) * 17




def _read_input() -> str:
    return (Path(__file__).with_name("test_data") / "day_15.txt").read_text()

def test_part_1_example_1():
    assert part_1(EXAMPLE_1) == 1320


def test_part_1_input():
    assert part_1(_read_input().rstrip()) == 517315


def test_part_2_example_1():
    assert part_2(EXAMPLE_1) == 145


def test_part_2_input():
    assert part_2(_read_input().rstrip()) == 247763


def test_hash():
    assert hash("HASH") == 52


EXAMPLE_1 = """rn=1,cm-,qp=3,cm=2,qp-,pc=4,ot=9,ab=5,pc-,pc=6,ot=7"""

File: rust/year_2023/src/day_15.rs
use std::array;

pub fn part_1(input: &str) -> usize {
    input.trim_end().split(',').map(hash).sum()
}

pub fn part_2(input: &str) -> usize {
    let mut hash_map = HashMap::new();
    for operation in input.trim_end().split(',') {
        if let Some(label) = operation.strip_suffix('-') {
            hash_map.remove(label);
        } else if let Some((label, focal_length)) = operation.split_once('=') {
            hash_map.insert(label, focal_length.parse().unwrap());
        } else {
            panic!("Invalid operation: {}", operation);
        }
    }
    return hash_map.focussing_power();
}

fn hash(input: &str) -> usize {
    input
        .bytes()
        .fold(0u8, |value, code| value.wrapping_add(code).wrapping_mul(17))
        .into()
}

#[derive(Debug)]
struct HashMap<'a> {
    buckets: [Vec<(&'a str, usize)>; 256],
}

impl<'a> HashMap<'a> {
    fn new() -> Self {
        Self {
            buckets: array::from_fn(|_| Vec::new()),
        }
    }

    fn insert(&mut self, key: &'a str, value: usize) {
        let i = hash(key);
        let bucket = &mut self.buckets[i];
        if let Some(j) = bucket.iter().position(|&(k, _)| k == key) {
            bucket[j] = (key, value);
        } else {
            bucket.push((key, value));
        }
    }

    fn remove(&mut self, key: &'a str) {
        let i = hash(key);
        let bucket = &mut self.buckets[i];
        if let Some(j) = bucket.iter().position(|&(k, _)| k == key) {
            bucket.remove(j);
        }
    }

    fn focussing_power(&self) -> usize {
        (1..)
            .zip(&self.buckets)
            .flat_map(|(i, bucket)| {
                (1..)
                    .zip(bucket)
                    .map(move |(j, (_, focal_length))| i * j * focal_length)
            })
            .sum()
    }
}

#[cfg(test)]
mod test {
    use test_case::test_case;

    const EXAMPLE_1: &str = "rn=1,cm-,qp=3,cm=2,qp-,pc=4,ot=9,ab=5,pc-,pc=6,ot=7";

    const INPUT: &str = include_str!("../../../testdata/year_2023/day_15.txt");

    #[test_case(EXAMPLE_1, 1320)]
    #[test_case(INPUT, 517315)]
    fn part_1(input: &str, expected: usize) {
        assert_eq!(super::part_1(input), expected);
    }

    #[test_case(EXAMPLE_1, 145)]
    #[test_case(INPUT, 247763)]
    fn part_2(input: &str, expected: usize) {
        assert_eq!(super::part_2(input), expected);
    }

    #[test_case("HASH", 52)]
    fn hash(input: &str, expected: usize) {
        assert_eq!(super::hash(input), expected);
    }
}
